tasty-plutus/src/Test/Tasty/Plutus/Internal/Monad.hs

```diff
@@ -0,0 +1,333 @@
+{-# LANGUAGE FunctionalDependencies #-}
+
+module Test.Tasty.Plutus.Internal.Monad (
+  MonadScriptTesting (
+    getShouldSucceed,
+    getMPred,
+    getCB,
+    getTestData,
+    getScript,
+    getOpts,
+    makeSuccess,
+    makeFailure
+  ),
+  SomeScript (SomeSpender, SomeMinter),
+  produceResult,
+) where
+
+import Control.Monad ((<=<))
+import Data.Functor ((<&>))
+import Data.Kind (Type)
+import Data.Text (Text)
+import Data.Vector (Vector)
+import Data.Vector qualified as Vector
+import Plutus.V1.Ledger.Api (
+  CurrencySymbol,
+  Datum (Datum),
+  Interval,
+  MintingPolicy,
+  POSIXTime,
+  Redeemer (Redeemer),
+  ScriptContext,
+  ToData (toBuiltinData),
+  TxId,
+  Validator,
+  ValidatorHash,
+  Value,
+ )
+import Plutus.V1.Ledger.Scripts (
+  Context (Context),
+  ScriptError (EvaluationError, EvaluationException, MalformedScript),
+ )
+import Plutus.V1.Pretty (scriptContextToValue)
+import Test.Tasty.Options (IsOption, OptionSet, lookupOption)
+import Test.Tasty.Plutus.Internal (
+  ContextBuilder,
+  Purpose (ForMinting, ForSpending),
+  ScriptResult (
+    InternalError,
+    NoOutcome,
+    ParseFailed,
+    ScriptFailed,
+    ScriptPassed
+  ),
+  TransactionConfig (TransactionConfig),
+  compileMinting,
+  compileSpending,
+  ourStyle,
+  testCurrencySymbol,
+  testFee,
+  testMintingPolicyScript,
+  testTimeRange,
+  testTxId,
+  testValidatorHash,
+  testValidatorScript,
+ )
+import Test.Tasty.Plutus.Options (
+  Fee (Fee),
+  PlutusTracing (Always, OnlyOnFail),
+  TestCurrencySymbol (TestCurrencySymbol),
+  TestTxId (TestTxId),
+  TestValidatorHash (TestValidatorHash),
+  TimeRange (TimeRange),
+ )
+import Test.Tasty.Plutus.TestData (Example (Bad, Good), TestData (MintingTest, SpendingTest))
+import Text.PrettyPrint (
+  Doc,
+  colon,
+  hang,
+  int,
+  renderStyle,
+  text,
+  vcat,
+  ($+$),
+  (<+>),
+  (<>),
+ )
+import Text.Show.Pretty (ppDoc, valToDoc)
+import Prelude hiding ((<>))
+
+data SomeScript (p :: Purpose) where
+  SomeSpender ::
+    Validator ->
+    SomeScript 'ForSpending
+  SomeMinter ::
+    MintingPolicy ->
+    SomeScript 'ForMinting
+
+class
```

Comment:
I get what this is going for, but I'm a bit skeptical. A lawless type class, even an internal one, is something I don't feel is warranted here, especially since there are only two instances.

Additionally, I feel this unifies _too_ much functionality, since it also handles construction of `Example`s, successes and failures. In reality, the main thing it should handle is the various environmental bits (`Context`, datum and/or redeemer, log) so that they can be 'called down' easily. 

I would say that `Reader conf a` is more than sufficient, with two different `conf` structures holding these bits of information, plus some helper functions.

Reply:
If I understood correctly, does that mean we should not provide do-it-all functions like `getScriptResult`, `handleError`, `deliverResult` etc here, but only the helper functions that provide `Doc`s for logging, plus utility functions like `getContext`?

Also, If we are to have different `conf` structures, will we still need to place some constraint on that, specifying how to extract the context, datum/redeemer, etc?

Reply:
Yes, the helpers should be producing `Doc`s, but the actual `conf` structures would probably only contain the datum, redeemer, `ScriptContext` and logs. The helpers could look like this:

```haskell
logToDoc :: forall (a :: Type) . (a -> [Text]) -> a -> Doc
logToDoc = -- whatever we currently do for log Docs
```

Then, we can use them in `Reader` by doing something like `asks (logToDoc getOutTheLog)`, where `getOutTheLog` would be specific to the `conf` type.